File: app/src/test/java/com/esafirm/androidboilerplate/DatabaseHelperTest.java
package com.esafirm.androidboilerplate;

import android.database.Cursor;
import com.esafirm.androidboilerplate.data.local.DatabaseHelper;
import com.esafirm.androidboilerplate.data.local.Db;
import com.esafirm.androidboilerplate.data.local.DbOpenHelper;
import com.esafirm.androidboilerplate.data.model.Ribot;
import com.esafirm.androidboilerplate.test.common.TestDataFactory;
import com.esafirm.androidboilerplate.util.DefaultConfig;
import com.esafirm.androidboilerplate.util.RxSchedulersOverrideRule;
import java.util.Arrays;
import java.util.List;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;
import rx.observers.TestSubscriber;

import static junit.framework.Assert.assertEquals;

/**
 * Unit tests integration with a SQLite Database using Robolectric
 */
@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, sdk = DefaultConfig.EMULATE_SDK)
public class DatabaseHelperTest {

  private final DatabaseHelper mDatabaseHelper =
      new DatabaseHelper(new DbOpenHelper(RuntimeEnvironment.application));

  @Rule public final RxSchedulersOverrideRule mOverrideSchedulersRule =
      new RxSchedulersOverrideRule();

  @Test public void setRibots() {
    Ribot ribot1 = TestDataFactory.makeRibot("r1");
    Ribot ribot2 = TestDataFactory.makeRibot("r2");
    List<Ribot> ribots = Arrays.asList(ribot1, ribot2);

    TestSubscriber<Ribot> result = new TestSubscriber<>();
    mDatabaseHelper.setRibots(ribots).subscribe(result);
    result.assertNoErrors();
    result.assertReceivedOnNext(ribots);

    Cursor cursor = mDatabaseHelper.getBriteDb()
        .query("SELECT * FROM " + Db.RibotProfileTable.INSTANCE.getTABLE_NAME());
    assertEquals(2, cursor.getCount());

    for (Ribot ribot : ribots) {
      cursor.moveToNext();
      assertEquals(ribot.getProfile(), Db.RibotProfileTable.INSTANCE.parseCursor(cursor));
    }
  }

  @Test public void getRibots() {
    Ribot ribot1 = TestDataFactory.makeRibot("r1");
    Ribot ribot2 = TestDataFactory.makeRibot("r2");
    List<Ribot> ribots = Arrays.asList(ribot1, ribot2);

    mDatabaseHelper.setRibots(ribots).subscribe();

    TestSubscriber<List<Ribot>> result = new TestSubscriber<>();
    mDatabaseHelper.getRibots().subscribe(result);
    result.assertNoErrors();
    result.assertValue(ribots);
  }
}

File: app/src/test/java/com/esafirm/androidboilerplate/DatabaseHelperTest.kt
package com.esafirm.androidboilerplate

import com.esafirm.androidboilerplate.data.local.DatabaseHelper
import com.esafirm.androidboilerplate.data.local.Db
import com.esafirm.androidboilerplate.data.local.DbOpenHelper
import com.esafirm.androidboilerplate.data.model.Ribot
import com.esafirm.androidboilerplate.test.common.TestDataFactory
import com.esafirm.androidboilerplate.util.DefaultConfig
import com.esafirm.androidboilerplate.util.RxSchedulersOverrideRule
import junit.framework.Assert.assertEquals
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.annotation.Config
import rx.observers.TestSubscriber
import java.util.*

/**
 * Unit tests integration with a SQLite Database using Robolectric
 */
@RunWith(RobolectricTestRunner::class)
@Config(constants = BuildConfig::class, sdk = intArrayOf(DefaultConfig.EMULATE_SDK))
class DatabaseHelperTest {

	private val mDatabaseHelper = DatabaseHelper(DbOpenHelper(RuntimeEnvironment.application))

	@JvmField @Rule val mOverrideSchedulersRule = RxSchedulersOverrideRule()

	@Test fun setRibots() {
		val ribot1 = TestDataFactory.makeRibot("r1")
		val ribot2 = TestDataFactory.makeRibot("r2")
		val ribots = Arrays.asList(ribot1, ribot2)

		val result = TestSubscriber<Ribot>()
		mDatabaseHelper.setRibots(ribots).subscribe(result)
		result.assertNoErrors()
		result.assertReceivedOnNext(ribots)

		val cursor = mDatabaseHelper.briteDb.query("SELECT * FROM " + Db.RibotProfileTable.TABLE_NAME)
		assertEquals(2, cursor.count)

		for (ribot in ribots) {
			cursor.moveToNext()
			assertEquals(ribot.profile, Db.RibotProfileTable.parseCursor(cursor))
		}
	}

	@Test fun getRibots() {
		val ribot1 = TestDataFactory.makeRibot("r1")
		val ribot2 = TestDataFactory.makeRibot("r2")
		val ribots = Arrays.asList(ribot1, ribot2)

		mDatabaseHelper.setRibots(ribots).subscribe()

		val result = TestSubscriber<List<Ribot>>()
		mDatabaseHelper.getRibots().subscribe(result)
		result.assertNoErrors()
		result.assertValue(ribots)
	}
}
